### PR TITLE
Add API Filename support to data classes, and check the Asset Manifes…

### PR DESCRIPTION
**Pull request by clocksmith**
…t for included fonts before http fetching

* Also adds example fonts and files for asset manifest manual testing in examples

NOTE: For review, may be easier to start with tests

### CODE REVIEW

File: lib/src/google_fonts_base.dart [hunk 1]
```diff
@@ -110,23 +114,33 @@ TextStyle googleFontsTextStyle({
...
+  if (_loadedFonts.contains(familyWithVariantString)) {
+    return;
+  }
+
+  // If this font can be loaded by the pre-bundled assets, then there is no
+  // need to load it at all.
+  final assetManifestJson = await _loadAssetManifestJson();
+
+  if (_isFamilyWithVariantInManifest(
+      descriptor.familyWithVariant, assetManifestJson)) {
```
Comment (johnsonmh) [inline lib/src/google_fonts_base.dart]: nit: add trailing comma and reformat.

```suggestion
      descriptor.familyWithVariant, assetManifestJson,)) {
```

File: lib/src/google_fonts_base.dart [hunk 2]
```diff
@@ -110,23 +114,33 @@ TextStyle googleFontsTextStyle({
...
   Future<ByteData> byteData;
   if (!kIsWeb) {
-    byteData = _readLocalFont(familyWithVariant);
+    byteData = _loadFontFromDeviceFileSystem(familyWithVariantString);
   }
   final localFontFound = byteData != null && await byteData != null;
   if (!localFontFound) {
-    byteData = _httpFetchFont(familyWithVariant, descriptor.fontUrl);
+    byteData = _httpFetchFontAndSaveToDevice(
+        familyWithVariantString, descriptor.fontUrl);
```
Comment (johnsonmh) [inline lib/src/google_fonts_base.dart]: nit: add trailing comma and reformat.

```suggestion
        familyWithVariantString, descriptor.fontUrl,);
```

File: lib/src/google_fonts_base.dart [hunk 3]
```diff
@@ -161,15 +175,16 @@ GoogleFontsVariant _closestMatch(
 /// it is the first time it is being loaded.
 ///
 /// This function can return null if the font fails to load from the URL.
-Future<ByteData> _httpFetchFont(String fontName, String fontUrl) async {
+Future<ByteData> _httpFetchFontAndSaveToDevice(
+    String fontName, String fontUrl) async {
```
Comment (johnsonmh) [inline lib/src/google_fonts_base.dart]: nit: trailing comma and reformat.

```suggestion
    String fontName, String fontUrl,) async {
```

File: lib/src/google_fonts_base.dart [hunk 4]
```diff
@@ -184,15 +199,16 @@ Future<String> get _localPath async {
 
 Future<File> _localFile(String name) async {
   final path = await _localPath;
+  // TODO(clocksmith): what if this file is an otf?
```
Comment (johnsonmh) [inline lib/src/google_fonts_base.dart]: Wow good call.

File: lib/src/google_fonts_base.dart [hunk 5]
```diff
@@ -221,3 +237,43 @@ int _computeMatch(GoogleFontsVariant a, GoogleFontsVariant b) {
...
+bool _isFamilyWithVariantInManifest(
+  GoogleFontsFamilyWithVariant familyWithVariant,
+  Map<String, dynamic> manifestJson,
+) {
+  if (manifestJson == null) return false;
+
+  for (final assetList in manifestJson.values) {
+    for (final String asset in assetList) {
+      String matchingFontSuffix;
+      for (final suffix in ['.ttf', '.otf']) {
```
Comment (johnsonmh) [inline lib/src/google_fonts_base.dart]: We should double check which font files are supported by fonts.google.com.

File: lib/src/google_fonts_family_with_variant.dart [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  const GoogleFontsFamilyWithVariant({
+    @required this.family,
+    @required this.googleFontsVariant,
+  })  : assert(family != null),
+        assert(googleFontsVariant != null);
+
+  final String family;
+  final GoogleFontsVariant googleFontsVariant;
+
+  GoogleFontsFamilyWithVariant.fromApiFilenamePrefix(String apiFilename)
```
Comment (johnsonmh) [inline lib/src/google_fonts_family_with_variant.dart]: I may be missing something, but I can't figure out where you actually use this constructor.

File: lib/src/google_fonts_base.dart (hunk 5 above)
Comment (clocksmith) [reply lib/src/google_fonts_base.dart]: I made a ticket for it

File: lib/src/google_fonts_family_with_variant.dart (hunk 6 above)
Comment (clocksmith) [reply lib/src/google_fonts_family_with_variant.dart]: You are right this is unused. I think I started out with it for symmetry with toApiFilenamePrefix, but it ended up not being needed

File: lib/src/google_fonts_variant.dart [hunk 7]
```diff
@@ -19,9 +19,58 @@ class GoogleFontsVariant {
...
+  }
+
+  static FontStyle _extractFontStyleFromApiFilenamePart(String filenamePart) {
+    if (filenamePart.contains('Italic')) return FontStyle.italic;
+    return FontStyle.normal;
+  }
+
+  String toApiFilenamePart() {
+    var weightPrefix;
+    if (fontWeight == FontWeight.w100) weightPrefix = 'Thin';
```
[the remark this answers is not in the document]
Comment (clocksmith) [reply lib/src/google_fonts_variant.dart]: I think this is totally reasonable. Done.

**Comment by clocksmith** [comment]
@googlebot I fixed it

**Comment by clocksmith** [comment]
new pull: https://github.com/material-foundation/google-fonts-flutter/pull/41

**Pull request closed without merging**